tools/tools.go

```diff
@@ -1,253 +1,295 @@
+		}
+
+		// for models where { or [ are used as tool calling
+		// tags, we only support parsing tools if the first non-
+		// whitespace character is { or [
+		if p.tag == "{" || p.tag == "[" {
+			if strings.TrimSpace(content) != "" {
+				p.state = toolsState_Done
+				return nil, content + string(p.buffer)
 			}
 		}
 
-		// Negative means we have an extra closing brace/bracket
-		if braceCount < 0 || squareCount < 0 {
-			return nil, errInvalidToolCall
+		if !found {
+			return nil, content
 		}
+
+		p.state = toolsState_ToolCalling
 	}
 
-	// If braces/brackets aren't balanced, need more input
-	if braceCount > 0 || squareCount > 0 {
-		return nil, errAccumulateMore
+	for {
+		call, end := p.findToolCall()
+		if call == nil {
+			break
+		}
+
+		call.Function.Index = p.n
+		p.n++
+		calls = append(calls, *call)
+
+		p.buffer = p.buffer[end:]
+		if len(p.buffer) == 0 {
+			break
+		}
 	}
 
-	t := strings.TrimSpace(s)
-	if len(t) == 0 {
-		return nil, errAccumulateMore
+	if p.done() {
+		p.state = toolsState_Done
+		content = string(p.buffer)
+		p.buffer = []byte{}
 	}
-	// If the input is a single square bracket, it's not a valid tool call
-	if t[0] == '[' && len(t) == 1 {
-		return nil, errAccumulateMore
+
+	return calls, content
+}
+
+// findTag processes a string to find and handle a tag pattern
+// returning true if the tag was found and false otherwise, and
+// a string content signaling any content that should be sent back to the user
+func (p *Parser) findTag() (int, bool) {
+	// First check for complete substring anywhere in s
+	if i := bytes.Index(p.buffer, []byte(p.tag)); i > -1 {
+		return i, true
 	}
 
-	// Attempt full unmarshal of the JSON
-	var toolCalls []api.ToolCall
-	for _, rawToolCall := range rawToolCalls {
-		var resp map[string]any
-		if err := json.Unmarshal([]byte(rawToolCall), &resp); err != nil {
-			continue
+	// Then check for partial suffix overlap
+	max := min(len(p.buffer), len(p.tag))
+	for i := max; i > 0; i-- {
+		if bytes.HasSuffix(p.buffer, []byte(p.tag[:i])) {
+			return len(p.buffer) - i, false
 		}
+	}
+	return -1, false
+}
 
-		// Collect nested objects that could contain tool calls
-		objs := collect(resp)
-		if len(objs) == 0 {
-			continue
-		}
+// findToolCall finds the next complete tool call in the given string
+// Returns the tool call, the number of characters consumed, and whether a tool call was found
+func (p *Parser) findToolCall() (*api.ToolCall, int) {
+	var name string
+	var args map[string]any
+	var end int = len(p.buffer)
 
-		// Extract tool calls from objects
-		for _, kv := range objs {
-			n, nok := kv[name].(string)
-			a, aok := kv[arguments].(map[string]any)
-			if nok && aok {
-				toolCalls = append(toolCalls, api.ToolCall{
-					Function: api.ToolCallFunction{
-						Name:      n,
-						Arguments: a,
-					},
-				})
-			} else {
-				slog.Debug("No valid tool call found in object.", "object", kv)
+	// find name
+	var i int
+	for _, n := range p.names {
```

Comment:
same idea here - `p.funcNames` or something similar to make it clear

Reply:
Given it's in the tool package and member of the tool parser – I think it should be relatively straightforward that it's the tool names?

Reply:
yeah not a big deal - just thought on first read it makes sense to differentiate between passed in tool names vs. parsed tool names. maybe `funcNames` was also not a good descriptor for that